### ISSUE TITLE
CRI: Clearly define the `Privileged` field in the security context

### ISSUE THREAD

**Issue by yujuhong** [kubernetes/kubernetes#44503]
CRI inherits the `Privileged` field from the kubernetes (and Docker) API. It's there to improve the user experience.
Since CRI is not intended to be directly user-facing, it should translate the field into a set of fields (e.g., capabilities).

At the least, we should clearly define what `Privileged` means in the CRI.

/cc @kubernetes/sig-node-api-reviews @Random-Liu @mrunalp @feiskyer @timstclair

**Comment by timstclair** [comment]
Alternatively, I think we should consider removing it. `privileged` is a docker concept, and isn't even part of the containerd API. To make it possible to remove privileged from the CRI, we would need to look at all the configuration docker sets with privileged, and make sure we provide those nobs in the CRI.

**Comment by yujuhong** [comment]
> Since CRI is not intended to be directly user-facing, it should translate the field into a set of fields (e.g., capabilities).

^^^ I did suggest removing it here. Sorry for not being clear.

**Comment by timstclair** [comment]
Ah, sorry for not reading carefully.

**Comment by Random-Liu** [comment]
> ^^^ I did suggest removing it here. Sorry for not being clear.

Agree. `Privileged` should still be supported in Kubernetes api, but should be removed from CRI.
The behavior should be defined and implemented in Kubelet.

**Comment by yujuhong** [comment]
> The behavior should be defined and implemented in Kubelet.

It needs tobe defined in the kubernetes API first :-)

**Comment by mrunalp** [comment]
SGTM 👍

**Comment by feiskyer** [comment]
LGTM.

It seems docker translate `Privileged` to a few oci configurations, including 

* Enable all capabilities
* Add all host devices
* Clear readonly on some volumes, e.g. cgroups and /sys volumes
* Clear security operations, e.g. apparmor

There is one concern of this: how could runtimes identify the container is privileged equivalent to docker without an explicitly `privileged` field?

**Comment by yujuhong** [comment]
> There is one concern of this: how could runtimes identify the container is privileged equivalent to docker without an explicitly privileged field?

They don't have to if kubelet simply translates "Privileged" to the set of configurations to the runtime. 

Another question is that if docker's privilege field changes, what should we do? That's why defining this at the kubernetes API level is important.

**Comment by feiskyer** [comment]
> They don't have to if kubelet simply translates "Privileged" to the set of configurations to the runtime.

Frakti and cri-o is depending on this to choose an alternative runtime, e.g. running them in docker while other regular containers in hypervisors.

**Comment by Random-Liu** [comment]
> Frakti and cri-o is depending on this to choose an alternative runtime, e.g. running them in docker while other regular containers in hypervisors.

Does cri-o also need this?

**Comment by mrunalp** [comment]
@Random-Liu Yes, we do.

**Comment by Random-Liu** [comment]
@mrunalp So cri-o does something special with `Previleged`? What special thing does cri-o do could not be expressed in CRI?

**Comment by timstclair** [comment]
If we removed privileged from the API, Frakti could instead make the decision on security options that it cannot support.  That said, we do need to think about what level of abstraction we want here. How do you handle other SecurityContext options, like seccomp, AppArmor, and capabilities in Frakti?

**Comment by mrunalp** [comment]
@Random-Liu cri-o supports mixed mode where regular containers use clear containers but privileged are run using runc. cc: @sameo

**Comment by feiskyer** [comment]
> How do you handle other SecurityContext options, like seccomp, AppArmor, and capabilities in Frakti?

Seccomp and AppArmor are not supported in frakti and containers in hypercontainer have full access to system operations (it is ok since they are inside hypervisors).

> If we removed privileged from the API, Frakti could instead make the decision on security options that it cannot support. That said, we do need to think about what level of abstraction we want here. 

Yep, that's the general idea.

**Comment by feiskyer** [comment]
> It seems docker translate Privileged to a few oci configurations, including
>
>* Enable all capabilities
>* Add all host devices
>* Clear readonly on some volumes, e.g. cgroups and /sys volumes
>* Clear security operations, e.g. apparmor

I think CRI could also translate in same way. 

@timstclair @Random-Liu @mrunalp Do you have any ideas on this?

**Comment by sameo** [comment]
@Random-Liu cri-o switches to a configured "privileged" runtime (typically `runc`) when either the sandbox security context `Privileged` field is on or the sandbox requires access to any host namespace (network, pid or ipc). Just for the record, it does not make sense for hypervisor based runtimes to support those options (You can't and you don't want to give a VM based container access to your host networking namespace) which is why we need to make such distinction. You may want to take that into account when implementing the containerd based CRI implementation unless you only intend to support namespace isolated containers.

I agree the `Privileged` semantics are not well defined and we currently assume it follows the implicit Docker ones. So I think it makes sense to remove it from the CRI and translate it into a well defined set of security knobs. We will translate those settings back from the CRI implementation to know which runtime we have to use.

On a similar topic, I think it would also make sense to add a user facing knob to the Kubernetes API for tagging pods as trusted or untrusted. It would be an opt-in setting and set to trusted by default, and would be used as a hint for CRI servers to make a decision on which runtime to pick or on which additional security mechanisms to apply for running unprivileged **and** untrusted sandboxes.

**Comment by egernst** [comment]
Making this more explicit is still desired.
